Minor rewrite of hashing function to improve performance by ~3x

danielweck opened the pull request:
Goober's current hashing function causes unnecessary memory allocations (`Array.split()`) as well as compute cycles / function calls (`Array.reduce()`). The following simple change does not increase the bundle size, and yields much better performance (at least 2x faster, nearly 3x):

```diff
-export let toHash = (str) =>
-    'go' + str.split('').reduce((out, i) => (101 * out + i.charCodeAt(0)) >>> 0, 11);
+export let toHash = (str) => {
+    let i = 0, l = str.length, out = 11;
+    while (i < l) out = (101 * out + str.charCodeAt(i++)) >>> 0
+    return 'go' + out;
+}
```

Comparisons using Benchmark.js:

`npm run test-perf-hash`

=>

```
> NODE_ENV=production node benchmarks/perf-hash.js


Starting HASH!
▸ goober original HASH x 3,180,521 ops/sec ±1.06% (87 runs sampled)
▸ goober optimized HASH x 8,219,240 ops/sec ±1.36% (84 runs sampled)

Fastest is: goober optimized HASH
```

```
> NODE_ENV=production node benchmarks/perf-hash.js


Starting HASH!
▸ goober original HASH x 3,582,669 ops/sec ±0.92% (90 runs sampled)
▸ goober optimized HASH x 9,067,875 ops/sec ±1.42% (94 runs sampled)

Fastest is: goober optimized HASH
```


```
> NODE_ENV=production node benchmarks/perf-hash.js

Starting HASH!
▸ goober original HASH x 3,517,773 ops/sec ±1.02% (89 runs sampled)
▸ goober optimized HASH x 10,189,264 ops/sec ±1.81% (90 runs sampled)

Fastest is: goober optimized HASH
```

```
> NODE_ENV=production node benchmarks/perf-hash.js

Starting HASH!
▸ goober original HASH x 3,368,369 ops/sec ±1.45% (85 runs sampled)
▸ goober optimized HASH x 9,699,551 ops/sec ±1.49% (90 runs sampled)

Fastest is: goober optimized HASH
```

danielweck commented:
Side note: you'll be pleased to know that Goober's hashing algorithm (pre-optimization) is already the fastest of the bunch ... naturally the downside is an increased risk of collision compared with other hashing schemes (and I suspect not as good distribution too). Do you have a reference / source for the technique used in Goober?

PS: based on my (very) basic tests, Twind's "cyrb32" implementation seems to hit the right spot:

https://gist.github.com/danielweck/fb48cafa2e6543e81f313e961ea8496c

CC @sastan

danielweck commented:
Update: I added a `Benchmark.js` suite in this PR:

`npm run test-perf-hash`

=>

```
> NODE_ENV=production node benchmarks/perf-hash.js


Starting HASH!
▸ goober original HASH x 3,180,521 ops/sec ±1.06% (87 runs sampled)
▸ goober optimized HASH x 8,219,240 ops/sec ±1.36% (84 runs sampled)

Fastest is: goober optimized HASH
```

cristianbote commented:
@danielweck this is freaking awesome! Love it! Thanks for the benchmark as well.

`goober`'s hash-ing algo was based on Java's string hash. It was/is the best size/perf ratio I could figure out at that moment. Of course, nothing is forever and everything is up for change. I remember talking with @jviide about it as well and he pointed out the same collision argument, IIRC.

Now, there's one other bench that I would like to run before settling on this, and that is the browser one 😃  cause these are running on node runtime. So, I went ahead and copied the above test in https://esbench.com/bench/605b608d6c89f600a57004d3 and oh my. You are right. This is by far the fastest.

The only other bench that doesn't get _that_ much perf boost is this one:
![Screenshot 2021-03-24 at 18 10 46](https://user-images.githubusercontent.com/3405161/112343936-46902180-8ccc-11eb-960e-7a3a8b35ca5e.png)

The size increase it is around `+16B` which I believe is worth it.

sastan commented:
It seems the twind version is faster. I do not know about the size impact.

![Bildschirmfoto 2021-03-24 um 18 37 54](https://user-images.githubusercontent.com/514405/112357783-219dad80-8cd0-11eb-8dc3-b3f5c35c4fe0.png)


https://esbench.com/bench/605b608d6c89f600a57004d3

danielweck commented:
Ah, just a word of caution, in your ESBench I noticed an error that comes from my own benchmark (but not in Goober's actual code, otherwise the unit tests wouldn't pass!)
See the change of place for `++` increment in this commit:

https://github.com/cristianbote/goober/pull/271/commits/479d367f53cd77ef01b07b0a5dd16b85014eca89

I re-run `npm run test-perf-hash` just to be sure :)

```
> NODE_ENV=production node benchmarks/perf-hash.js

Starting HASH!
▸ goober original HASH x 3,517,773 ops/sec ±1.02% (89 runs sampled)
▸ goober optimized HASH x 10,189,264 ops/sec ±1.81% (90 runs sampled)

Fastest is: goober optimized HASH
```

```
> NODE_ENV=production node benchmarks/perf-hash.js

Starting HASH!
▸ goober original HASH x 3,368,369 ops/sec ±1.45% (85 runs sampled)
▸ goober optimized HASH x 9,699,551 ops/sec ±1.49% (90 runs sampled)

Fastest is: goober optimized HASH
```

danielweck commented:
> It seems the twind version is faster.

Strange ... I added Twind's hash routine to the benchmark, it is much slower:

# Goober original
```js
    const c = 'go' + str.split('').reduce((out, s) => (101 * out + s.charCodeAt(0)) >>> 0, 11);
```

# Goober optimized
```js
    let i = 0, l = str.length, out = 11;
    while (i < l) out = (101 * out + str.charCodeAt(i++)) >>> 0
    const c = 'go' + out;
```

# Twind (`cyrb32` with `Math.imul()`)
```js
    let i = str.length - 1, out = 9;
    while (i >= 0) out = Math.imul(out ^ str.charCodeAt(i--), 0x5f356495);
    const c = 'go' + ((out ^ (out >>> 9)) >>> 0).toString(36);
```

# RESULTS

`npm run test-perf-hash`
=>
```
> NODE_ENV=production node benchmarks/perf-hash.js

Starting HASH!
▸ goober original HASH x 3,512,636 ops/sec ±1.03% (91 runs sampled)
▸ goober optimized HASH x 10,347,402 ops/sec ±1.02% (90 runs sampled)
▸ twind HASH x 2,458,171 ops/sec ±0.70% (96 runs sampled)

Fastest is: goober optimized HASH
```
```
> NODE_ENV=production node benchmarks/perf-hash.js


Starting HASH!
▸ goober original HASH x 3,464,524 ops/sec ±1.61% (85 runs sampled)
▸ goober optimized HASH x 10,256,783 ops/sec ±1.07% (92 runs sampled)
▸ twind HASH x 2,453,381 ops/sec ±0.80% (94 runs sampled)

Fastest is: goober optimized HASH
```

danielweck commented:
Wow, ESBench running in the browser (as opposed to Benchmark.js running in Node) => HUUUGE difference: Twind wins in this case! (as @sastan said) I guess `Math.imul()` is not optimized in NodeJS? (note that I am using Firefox)

<img width="443" alt="Screenshot 2021-03-24 at 19 04 58" src="https://user-images.githubusercontent.com/625201/112369278-eac99680-8cd3-11eb-9d38-536ceef768ab.png">


```
const STR = `abcdefghijklmnopqrstuvwyxz ABCDEFGHIJKLMNOPQRSTUVWYXZ 0123456789`;

const original = (str) => {
    const c = 'go' + str.split('').reduce((out, s) => (101 * out + s.charCodeAt(0)) >>> 0, 11);
};

const optimized = (str) => {
    let i = 0, l = str.length, out = 11;
    while (i < l) out = (101 * out + str.charCodeAt(i++)) >>> 0
    const c = 'go' + out;
};

const twind = (str) => {
    let i = str.length - 1, out = 9;
    while (i >= 0) out = Math.imul(out ^ str.charCodeAt(i--), 0x5f356495);
    const c = 'go' + ((out ^ (out >>> 9)) >>> 0).toString(36);
};
```

sastan commented:
That is interesting! I run that in Safari. Re-benching in Chrome confirm your results.

![Bildschirmfoto 2021-03-24 um 20 09 54](https://user-images.githubusercontent.com/514405/112369799-e786d880-8cdc-11eb-983e-17af08991215.png)

What we can say for sure is that the optimized version is faster than the current goober one.

danielweck commented:
Safari same as Firefox (Twind / `Math.imul()` / `cyrb32` much faster)

<img width="563" alt="Screenshot 2021-03-24 at 19 09 46" src="https://user-images.githubusercontent.com/625201/112369989-c4582b00-8cd4-11eb-95c7-6ae378f1170c.png">

danielweck commented:
...BUT, in Chrome Twind's algo is slow!

Go figure!?

<img width="495" alt="Screenshot 2021-03-24 at 19 12 09" src="https://user-images.githubusercontent.com/625201/112370153-f2d60600-8cd4-11eb-883b-f4de5635b525.png">

<img width="495" alt="Screenshot 2021-03-24 at 19 12 38" src="https://user-images.githubusercontent.com/625201/112370157-f49fc980-8cd4-11eb-9b1d-a1c854d7aeed.png">

sastan commented:
Using [perf.link](https://perf.link/#eyJpZCI6ImhucmZoaGsyZGUwIiwidGl0bGUiOiJnb29iZXItaGFzaC1leHBsb3JhdGlvbiIsImJlZm9yZSI6ImNvbnN0IFNUUiA9IGBhYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3l4eiBBQkNERUZHSElKS0xNTk9QUVJTVFVWV1lYWiAwMTIzNDU2Nzg5YDtcblxuY29uc3Qgb3JpZ2luYWwgPSAoc3RyKSA9PiB7XG4gIGNvbnN0IGMgPSAnZ28nICsgc3RyLnNwbGl0KCcnKS5yZWR1Y2UoKG91dCwgcykgPT4gKDEwMSAqIG91dCArIHMuY2hhckNvZGVBdCgwKSkgPj4%2BIDAsIDExKTtcbn07XG5cbmNvbnN0IG9wdGltaXplZCA9IChzdHIpID0%2BIHtcbiAgICBsZXQgaSA9IDAsIGwgPSBzdHIubGVuZ3RoIC0gMSwgdiA9IDExO1xuICAgIHdoaWxlIChpKysgPCBsKSB2ID0gKDEwMSAqIHYgKyBzdHIuY2hhckNvZGVBdChpKSkgPj4%2BIDBcbiAgICBjb25zdCBjID0gJ2dvJyArIHY7XG59O1xuXG5jb25zdCBjeXJiMzIgPSAodmFsdWUpID0%2BIHtcbiAgZm9yICh2YXIgaCA9IDksIGluZGV4ID0gdmFsdWUubGVuZ3RoOyBpbmRleC0tOyApIHtcbiAgICBoID0gTWF0aC5pbXVsKGggXiB2YWx1ZS5jaGFyQ29kZUF0KGluZGV4KSwgMHg1ZjM1NjQ5NSlcbiAgfVxuXG4gIHJldHVybiAndHctJyArICgoaCBeIChoID4%2BPiA5KSkgPj4%2BIDApLnRvU3RyaW5nKDM2KVxufSIsInRlc3RzIjpbeyJuYW1lIjoib3JpZ2luYWwiLCJjb2RlIjoib3JpZ2luYWwoU1RSKTsiLCJydW5zIjpbMjUwMDAsNTcwMDAsNzkwMDAsMTAwMCw2ODAwMCwyODAwMCwzNTAwMCw3NzAwMCw3NDAwMCwxMDAwLDEyMDAwLDYyMDAwLDEwMDAsNTEwMDAsMTAwMCw1MDAwMCw4MDAwLDMxMDAwLDQ2MDAwLDI3MDAwLDEwMDAsMTAwMCwxMDAwLDQwMDAsMjAwMDAsMzYwMDAsMTAwMCwzNDAwMCw1NTAwMCwxMDAwLDM4MDAwLDUwMDAsOTcwMDAsMjgwMDAsOTEwMDAsMTAwMCw3NTAwMCw3MTAwMCwxMDAwLDEwMDAsNjAwMCwxMDAwLDEwMDAsMTMwMDAsMTAwMCw4MDAwLDE3MDAwLDYzMDAwLDM2MDAwLDQxMDAwLDYwMDAsMzAwMDAsMTAwMCwxMzAwMCwyMzAwMCwyNDAwMCwzNDAwMCwxNzAwMCw5NzAwMCwxNzAwMCw0NzAwMCwxNjAwMCw0MDAwLDAsMzIwMDAsMjAwMCwzNjAwMCw0NDAwMCwxMTkwMDAsMzMwMDAsNTAwMCw0MzAwMCw4MDAwLDQzMDAwLDEwMDAsMTQwMDAsMjAwMCwxMDAwLDEwMDAsNzAwMCwxMDAwLDMzMDAwLDYzMDAwLDEwMDAsODAwMDAsNzYwMDAsNDgwMDAsNTgwMDAsOTcwMDAsMTAwMCwxODAwMCwxMDAwLDEwMDAsMjAwMCwxMDAwLDEwMDAsMzUwMDAsMTAwMCwxMDAwLDI2MDAwXSwib3BzIjoyNzQ4MH0seyJuYW1lIjoib3B0aW1pemVkIiwiY29kZSI6Im9wdGltaXplZChTVFIpOyIsInJ1bnMiOlsxMzIwMDAsMjM0MDAwLDI0NTAwMCwxNjAwMCwyNTgwMDAsMTI0MDAwLDE4NTAwMCwyNjEwMDAsMjcyMDAwLDQwMDAsODcwMDAsMTk1MDAwLDIxMDAwLDIwMTAwMCwxNzMwMDAsMTgxMDAwLDc1MDAwLDEzNjAwMCwxNzgwMDAsMTIzMDAwLDEwMDAwLDUwMDAsMjIwMDAsNDQwMDAsMTAwMCwxODEwMDAsNTUwMDAsMTgyMDAwLDIwNzAwMCwzNzAwMCwxNTUwMDAsNDUwMDAsMjcxMDAwLDE0MjAwMCwyNTQwMDAsMTAwMCwyNTYwMDAsMjA4MDAwLDU5MDAwLDIzMDAwLDM4MDAwLDQwMDAsNTQwMDAsOTIwMDAsNzAwMCw1NjAwMCwxMTQwMDAsMjE3MDAwLDE2MjAwMCwxNDgwMDAsNTYwMDAsMTAwMDAwLDE3MDAwLDEwMTAwMCw5NjAwMCwxMzMwMDAsMTE2MDAwLDEwNjAwMCwzMDIwMDAsMTA1MDAwLDE5MDAwMCw4NDAwMCw0NTAwMCwxMzgwMDAsMTMxMDAwLDM1MDAwLDEzOTAwMCwxODQwMDAsMjk0MDAwLDE1ODAwMCw0MzAwMCwxNzQwMDAsNzEwMDAsMTg4MDAwLDE4MDAwLDc0MDAwLDY0MDAwLDIzNzAwMCwzMzAwMCw1NDAwMCwxMTAwMCwxNDIwMDAsMjMxMDAwLDExMDAwLDI1MjAwMCwxMDAwLDE3NzAwMCwyNDQwMDAsMjU3MDAwLDEwMDAsODQwMDAsMzAwMCwxMTAwMCwyNzAwMCw0MDAwLDIzMjAwMCwxNTUwMDAsMzQwMDAsMjIwMDAsNzIwMDBdLCJvcHMiOjExNjA4MH0seyJuYW1lIjoiY3lyYjMyIiwiY29kZSI6ImN5cmIzMihTVFIpIiwicnVucyI6WzI0MTAwMCw1MDEwMDAsNDk2MDAwLDEwMDAsNTg1MDAwLDI2NDAwMCw0NDYwMDAsNTExMDAwLDY4MzAwMCwxMDAwLDE0MjAwMCw0MjIwMDAsNzAwMCwzOTAwMDAsMzUxMDAwLDM5MTAwMCwxMTUwMDAsMjI2MDAwLDM4MjAwMCwyMjYwMDAsMjExMDAwLDEwMDAsMTAwMCwzODAwMCwyNDcwMDAsMzgxMDAwLDE1MDAwLDQwMTAwMCw0MzgwMDAsMTcwMDAsMzAwMDAwLDYzMDAwLDUyNzAwMCwyNTIwMDAsNTU5MDAwLDEwMDAsNjAwMDAwLDQ5NTAwMCw4NDAwMCwxNzAwMCw0NDAwMCwxMDAwLDEwMDAsMTE3MDAwLDIxMDAwLDcxMDAwLDIxMTAwMCw0OTgwMDAsMzU0MDAwLDM2MDAwMCw0NzAwMCwyMTYwMDAsMTQwMDAsOTgwMDAsMTU5MDAwLDE5OTAwMCwzNjIwMDAsMTgyMDAwLDEwNzUwMDAsMTk0MDAwLDM5MjAwMCwxMzUwMDAsMjcwMDAsMjI5MDAwLDIzMTAwMCwzMTAwMCwzMTYwMDAsNDE0MDAwLDYwNjAwMCwyMjQwMDAsNTQwMDAsMzc5MDAwLDcwMDAwLDQwMTAwMCw0MzAwMCwxMzUwMDAsMzQwMDAsMTAwMCwyNzAwMCw3NzAwMCw3MDAwLDI4NzAwMCw1MTAwMDAsMjAwMCw1MjkwMDAsMTAwMCwzOTgwMDAsNTA5MDAwLDY1MjAwMCwxMDAwLDE1MDAwMCwxMDAwLDYwMDAsMTUwMDAsMTAwMCwxMDAwLDMzNzAwMCwxMTAwMCwyMTAwMCwxMTgwMDBdLCJvcHMiOjIyNjM2MH1dLCJ1cGRhdGVkIjoiMjAyMS0wMy0yNFQxOToxMjoyNC42MDlaIn0%3D) 

- Safari: same results
- Chrome: optimized and cyrb32 almost the same with advantages for optimized

sastan commented:
Important for gobber: The optimized version is faster than the current in all browsers

danielweck commented:
@sastan you are using code snippets with my copy-paste error, please use this from now on (sorry):

```js
const STR = `abcdefghijklmnopqrstuvwyxz ABCDEFGHIJKLMNOPQRSTUVWYXZ 0123456789`;

const original = (str) => {
    const c = 'go' + str.split('').reduce((out, s) => (101 * out + s.charCodeAt(0)) >>> 0, 11);
};

const optimized = (str) => {
    let i = 0, l = str.length, out = 11;
    while (i < l) out = (101 * out + str.charCodeAt(i++)) >>> 0
    const c = 'go' + out;
};

const twind = (str) => {
    let i = str.length - 1, out = 9;
    while (i >= 0) out = Math.imul(out ^ str.charCodeAt(i--), 0x5f356495);
    const c = 'go' + ((out ^ (out >>> 9)) >>> 0).toString(36);
};
```

sastan commented:
Here is the updated [perf.link](https://perf.link/#eyJpZCI6ImhucmZoaGsyZGUwIiwidGl0bGUiOiJnb29iZXItaGFzaC1leHBsb3JhdGlvbiIsImJlZm9yZSI6ImNvbnN0IFNUUiA9IGBhYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3l4eiBBQkNERUZHSElKS0xNTk9QUVJTVFVWV1lYWiAwMTIzNDU2Nzg5YDtcblxuY29uc3Qgb3JpZ2luYWwgPSAoc3RyKSA9PiB7XG4gICAgY29uc3QgYyA9ICdnbycgKyBzdHIuc3BsaXQoJycpLnJlZHVjZSgob3V0LCBzKSA9PiAoMTAxICogb3V0ICsgcy5jaGFyQ29kZUF0KDApKSA%2BPj4gMCwgMTEpO1xufTtcblxuY29uc3Qgb3B0aW1pemVkID0gKHN0cikgPT4ge1xuICAgIGxldCBpID0gMCwgbCA9IHN0ci5sZW5ndGgsIG91dCA9IDExO1xuICAgIHdoaWxlIChpIDwgbCkgb3V0ID0gKDEwMSAqIG91dCArIHN0ci5jaGFyQ29kZUF0KGkrKykpID4%2BPiAwXG4gICAgY29uc3QgYyA9ICdnbycgKyBvdXQ7XG59O1xuXG5jb25zdCBjeXJiMzIgPSAodmFsdWUpID0%2BIHtcbiAgZm9yICh2YXIgaCA9IDksIGluZGV4ID0gdmFsdWUubGVuZ3RoOyBpbmRleC0tOyApIHtcbiAgICBoID0gTWF0aC5pbXVsKGggXiB2YWx1ZS5jaGFyQ29kZUF0KGluZGV4KSwgMHg1ZjM1NjQ5NSlcbiAgfVxuXG4gIHJldHVybiAndHctJyArICgoaCBeIChoID4%2BPiA5KSkgPj4%2BIDApLnRvU3RyaW5nKDM2KVxufSIsInRlc3RzIjpbeyJuYW1lIjoib3JpZ2luYWwiLCJjb2RlIjoib3JpZ2luYWwoU1RSKTsiLCJydW5zIjpbMzUwMDAsNzQwMDAsMTAwMCwxMDAwLDI3MDAwLDEwMDAsNzcwMDAsNDYwMDAsNTEwMDAsNjQwMDAsNjAwMCw0MTAwMCw3MzAwMCwxMDAwLDEwMDAsMjAwMCwzNjAwMCwxMDAwLDEwMDAsNzAwMCwyMTAwMCw4NTAwMCwxMDAwLDQxMDAwLDYyMDAwLDQxMDAwLDUzMDAwLDYwMDAsMTAwMCw0MDAwLDIwMDAsNDkwMDAsODAwMCwxMDAwLDUzMDAwLDEyMDAwLDcyMDAwLDEwMDAsNDYwMDAsMTAwMCwzMDAwLDQ5MDAwLDYwMDAsMTAwMCwxMDAwLDEwMDAsMTIwMDAsNDAwMDAsMzYwMDAsMTAwMCw3MTAwMCwyMDAwLDEwMDAsMjAwMDAsMTgwMDAsNDkwMDAsOTAwMCw0MjAwMCwxMDAwLDIwMDAsMzYwMDAsMjAwMCwxMDAwLDEwMDAsMTAwMCwxMDAwLDcwMDAsMTAwMCwzNDAwMCwxMDAwLDEwMDAsMTAwMCw3NDAwMCw3MTAwMCw0MDAwLDIwMDAsMTAwMCwxMDAwLDE2MDAwLDM0MDAwLDUwMDAsMTAwMCw3NjAwMCw0MjAwMCwyMjAwMCw0NTAwMCwzMDAwLDgwMDAsMTAwMCw3MjAwMCw3NjAwMCw0ODAwMCw2ODAwMCwxMDAwLDUxMDAwLDgwMDAsNDIwMDAsMTAwMCw0MjAwMCw3MDAwMF0sIm9wcyI6MjQ0MzB9LHsibmFtZSI6Im9wdGltaXplZCIsImNvZGUiOiJvcHRpbWl6ZWQoU1RSKTsiLCJydW5zIjpbMTUzMDAwLDEwMDAsMTAwMCwxMzAwMCwxMzAwMDAsNzAwMCwyMzAwMDAsMjM4MDAwLDIxNDAwMCwyMjkwMDAsODAwMDAsMTYyMDAwLDE4NjAwMCwzMDAwLDE5MDAwMCw1ODAwMCwxNjEwMDAsNzAwMCwxNTAwMCwxOTAwMCwxMDQwMDAsMjY3MDAwLDM1MDAwLDEwMDAsMjQ3MDAwLDIwODAwMCwyMTcwMDAsMzgwMDAsMjAwMCw1MDAwMCwzMTAwMCwxODUwMDAsNTYwMDAsMzIwMDAsMjEzMDAwLDc4MDAwLDI0MjAwMCwyOTAwMCwxODUwMDAsMTIwMDAsMjIwMDAsMTc2MDAwLDQ1MDAwLDIzMDAwLDcxMDAwLDkwMDAsOTIwMDAsMTY0MDAwLDE1OTAwMCwxMDAwLDIzNTAwMCw1NTAwMCwxNTAwMCwxMTMwMDAsOTEwMDAsMTg4MDAwLDYzMDAwLDE0NTAwMCwxMDAwLDUwMDAwLDEzODAwMCwzMjAwMCwxMDAwLDEyMDAwLDgwMDAsMTYwMDAsNjYwMDAsMjIwMDAsMTI2MDAwLDExOTAwMCwxNTAwMCw4MDAwLDI1NDAwMCwyNDcwMDAsNDkwMDAsMTgwMDAsMzcwMDAsMTAwMCw4NzAwMCwxMzgwMDAsNDUwMDAsMTcwMDAsMjcxMDAwLDE3MzAwMCwxMDAwMDAsMTY3MDAwLDc1MDAwLDUzMDAwLDEwMDAsMjU0MDAwLDI0OTAwMCwxOTcwMDAsMjQwMDAwLDIxMDAwLDIzNzAwMCw1OTAwMCwxMzcwMDAsMTQxMDAwLDE3NDAwMCwyNDYwMDBdLCJvcHMiOjEwMjk4MH0seyJuYW1lIjoiY3lyYjMyIiwiY29kZSI6ImN5cmIzMihTVFIpIiwicnVucyI6WzM1NTAwMCw0MjkwMDAsMTAwMCw3MDAwLDIzMTAwMCwxMDAwLDQyNDAwMCw0NzEwMDAsNDczMDAwLDQ5NzAwMCwxMjcwMDAsMzYxMDAwLDM3MTAwMCwxMDAwLDEwMDAsODQwMDAsMzcwMDAwLDQwMDAsNzAwMCw0NzAwMCwxNzYwMDAsNjQ0MDAwLDMxMDAwLDEwMDAsNTE3MDAwLDQ1MjAwMCw0NjkwMDAsNDcwMDAsNzAwMCw2NTAwMCwyNzAwMCw0MDIwMDAsOTIwMDAsMjYwMDAsNDcxMDAwLDEyNDAwMCw1MDkwMDAsMjcwMDAsMzQ4MDAwLDEwMDAwLDYwMDAwLDM4MzAwMCwxMDgwMDAsMTkwMDAsNjkwMDAsNTAwMCw2OTAwMCwzNjUwMDAsMzc4MDAwLDQzNzAwMCw1MTUwMDAsMzIwMDAsMjEwMDAsMTc3MDAwLDEzODAwMCw0MTkwMDAsNzIwMDAsMzU3MDAwLDEwMDAsNjQwMDAsMjMxMDAwLDMxMDAwLDEwMDAsNzAwMCwyMDAwLDI5MDAwLDk3MDAwLDE0MDAwLDI1OTAwMCwzMjgwMDAsNzAwMCwxMDAwLDYzNjAwMCw1OTcwMDAsMzkwMDAsMjMwMDAsMjQwMDAsMTAwMCwxMjMwMDAsMjk0MDAwLDQ2MDAwLDcwMDAsNTU0MDAwLDQwMzAwMCwxODAwMDAsMzEwMDAwLDEyMzAwMCwxMDIwMDAsMTAwMCw1OTMwMDAsNjI0MDAwLDQxMTAwMCw0ODMwMDAsOTAwMCw0MTUwMDAsOTYwMDAsMzY2MDAwLDEwMDAsMzAyMDAwLDUwMjAwMF0sIm9wcyI6MjA2NjgwfV0sInVwZGF0ZWQiOiIyMDIxLTAzLTI0VDE5OjE3OjMzLjE1NFoifQ%3D%3D)

Safari: cyrb32 (206,680 ops/s), optimized (102,980 ops/s) 
Chrome: very different results original and optimized almost the same, then cyrb32

danielweck commented:
I'm on MacOS, I tried [perf.link](https://perf.link/#eyJpZCI6ImhucmZoaGsyZGUwIiwidGl0bGUiOiJnb29iZXItaGFzaC1leHBsb3JhdGlvbiIsImJlZm9yZSI6ImNvbnN0IFNUUiA9IGBhYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3l4eiBBQkNERUZHSElKS0xNTk9QUVJTVFVWV1lYWiAwMTIzNDU2Nzg5YDtcblxuY29uc3Qgb3JpZ2luYWwgPSAoc3RyKSA9PiB7XG4gICAgY29uc3QgYyA9ICdnbycgKyBzdHIuc3BsaXQoJycpLnJlZHVjZSgob3V0LCBzKSA9PiAoMTAxICogb3V0ICsgcy5jaGFyQ29kZUF0KDApKSA%2BPj4gMCwgMTEpO1xufTtcblxuY29uc3Qgb3B0aW1pemVkID0gKHN0cikgPT4ge1xuICAgIGxldCBpID0gMCwgbCA9IHN0ci5sZW5ndGgsIG91dCA9IDExO1xuICAgIHdoaWxlIChpIDwgbCkgb3V0ID0gKDEwMSAqIG91dCArIHN0ci5jaGFyQ29kZUF0KGkrKykpID4%2BPiAwXG4gICAgY29uc3QgYyA9ICdnbycgKyBvdXQ7XG59O1xuXG5jb25zdCB0d2luZCA9IChzdHIpID0%2BIHtcbiAgICBsZXQgaSA9IHN0ci5sZW5ndGggLSAxLCBvdXQgPSA5O1xuICAgIHdoaWxlIChpID49IDApIG91dCA9IE1hdGguaW11bChvdXQgXiBzdHIuY2hhckNvZGVBdChpLS0pLCAweDVmMzU2NDk1KTtcbiAgICBjb25zdCBjID0gJ2dvJyArICgob3V0IF4gKG91dCA%2BPj4gOSkpID4%2BPiAwKS50b1N0cmluZygzNik7XG59OyIsInRlc3RzIjpbeyJuYW1lIjoib3JpZ2luYWwiLCJjb2RlIjoib3JpZ2luYWwoU1RSKTsiLCJydW5zIjpbOTAwMDAsMTA3MDAwLDQwMDAwLDEwMDAsNjgwMDAsMTAwMCw2MDAwLDgwMDAwLDEwNTAwMCwxMDUwMDAsOTAwMDAsODAwMCwxMDAwLDgyMDAwLDE1MDAwMCw3MjAwMCwxMTIwMDAsMTYxMDAwLDUzMDAwLDEyMjAwMCw5MzAwMCw3MDAwLDMyMDAwLDExMzAwMCwyMDAwMCwxMDAwLDEzMzAwMCwxMTcwMDAsNDAwMDAsOTkwMDAsMTA3MDAwLDYwMDAsMTI0MDAwLDQ2MDAwLDU2MDAwLDcwMDAwLDE2MDAwLDEwMDAsNzYwMDAsMjUwMDAsMjYwMDAsMTU4MDAwLDM3MDAwLDE2NjAwMCw0MDAwLDI3MDAwLDEwMDAsNzAwMDAsNzQwMDAsMTMyMDAwLDkzMDAwLDExNTAwMCwxMTAwMDAsNTgwMDAsMzcwMDAsMjcwMDAsMzEwMDAsMTY2MDAwLDc2MDAwLDYyMDAwLDg3MDAwLDk0MDAwLDcxMDAwLDEwMDAwMCw2NTAwMCw3NDAwMCwxMzAwMDAsOTkwMDAsNTMwMDAsMTU1MDAwLDEwMDAsNjgwMDAsODEwMDAsNDEwMDAsMTExMDAwLDg2MDAwLDEwMDAsMTY2MDAwLDUzMDAwLDg5MDAwLDIxMDAwLDE2MDAwLDE3MDAwLDYwMDAsNDEwMDAsMzcwMDAsODQwMDAsNzQwMDAsMjIwMDAsMzMwMDAsOTkwMDAsMTUxMDAwLDQwMDAsMTE1MDAwLDU2MDAwLDEwMDAsNTAwMDAsNjQwMDAsNjAwMDAsOTcwMDBdLCJvcHMiOjY3NzkwfSx7Im5hbWUiOiJvcHRpbWl6ZWQiLCJjb2RlIjoib3B0aW1pemVkKFNUUik7IiwicnVucyI6WzU0MDAwLDEzMzAwMCwxNzAwMCwxNTEwMDAsNjAwMCw2OTAwMCw1NTAwMCw1MjAwMCw5MjAwMCw1NTAwMCwxMDEwMDAsNTQwMDAsOTkwMDAsMTAwMCwxMDUwMDAsMzkwMDAsMTAxMDAwLDEwMTAwMCw2MzAwMCw0NDAwMCwxMDEwMDAsMTAxMDAwLDEwMDAsNDAwMDAsMjMwMDAsNjAwMDAsMTAxMDAwLDM4MDAwLDUxMDAwLDE0MTAwMCw3MjAwMCwxNzAwMCw3ODAwMCwxMDAwLDkwMDAwLDcyMDAwLDEyMjAwMCwxMTMwMDAsNjAwMCw0MjAwMCw1MDAwMCw5MDAwLDg3MDAwLDUwMDAsMzEwMDAsMzAwMCwxMTIwMDAsMTAxMDAwLDIwMDAwLDEwOTAwMCw1MTAwMCwxMDEwMDAsNjEwMDAsNDMwMDAsNjQwMDAsOTQwMDAsMTA3MDAwLDE1MTAwMCwyMzAwMCwyMDAwMCwzNDAwMCw1MDAwMCwxMDEwMDAsNDQwMDAsMTQwMDAsMzYwMDAsOTkwMDAsNTMwMDAsMTEwMDAwLDU3MDAwLDEwMjAwMCwxMzEwMDAsNzIwMDAsNDEwMDAsODQwMDAsMTMwMDAsMTAwMCw1MDAwMCwzNjAwMCw3MTAwMCwzMDAwMCwzMzAwMCwxNjAwMCwxMDEwMDAsMTAxMDAwLDEwMTAwMCwxMDEwMDAsODAwMCwyNzAwMCwxMDAwLDMwMDAwLDEwMjAwMCw0MDAwLDY3MDAwLDY0MDAwLDEwMDAsMjAwMCwxNTcwMDAsNzEwMDAsNDYwMDBdLCJvcHMiOjYxNjYwfSx7Im5hbWUiOiJjeXJiMzIiLCJjb2RlIjoidHdpbmQoU1RSKSIsInJ1bnMiOlsxMTAwMCw3MzAwMCwxMDAwLDEyNDAwMCwxMDIwMDAsNjUwMDAsMTUwMDAsNjAwMDAsOTAwMCwxMTYwMDAsMTAwMCwxMTkwMDAsOTYwMDAsMjAwMDAsMTcwMDAsMTIwMDAsNTcwMDAsOTcwMDAsNjAwMCwxMDAwMCw5NTAwMCwzMzAwMCw4ODAwMCwxMDEwMDAsMTAwMCw1MTAwMCwxMDYwMDAsMCwxMDEwMDAsMTE5MDAwLDgxMDAwLDE3MDAwLDgyMDAwLDYzMDAwLDc1MDAwLDg3MDAwLDExOTAwMCwxMTkwMDAsMTAwMCwxMTkwMDAsMTE5MDAwLDExOTAwMCw0MTAwMCwxMDAwLDQwMDAwLDEwMDAsNjAwMDAsMzcwMDAsODYwMDAsMTA3MDAwLDI1MDAwLDE5MDAwLDYwMDAsMTAwMCw0NTAwMCwzODAwMCwzMzAwMCwxMTEwMDAsMTAwMCwxMDAwLDQwMDAsNjUwMDAsNjAwMCw0NzAwMCw2MDAwLDEwMDAsMzAwMCw4MDAwLDE2MDAwLDExOTAwMCw1MDAwLDEwMDAsMzIwMDAsMTIwMDAsNzgwMDAsNDIwMDAsNTQwMDAsNTQwMDAsMTAwMCw4NzAwMCwxMDAwLDEyMDAwLDYwMDAsMTQwMDAsMTAwMCwxMTkwMDAsMTEwMDAwLDExMDAwLDMxMDAwLDQwMDAwLDQwMDAwLDcyMDAwLDQ4MDAwLDI0MDAwLDE0MDAwLDYwMDAwLDg2MDAwLDY4MDAwLDM0MDAwLDEwMDBdLCJvcHMiOjQ3OTMwfV0sInVwZGF0ZWQiOiIyMDIxLTAzLTI0VDE5OjE5OjA4Ljg4OVoifQ%3D%3D) (thanks @sastan ):

Both Safari and Firefox report that Twind / `cyrb32` is faster. Chrome is the odd one out, where the original and optimized Goober hash routines achieve similar perfs, in fact the original is faster in most my runs!! :) (the pitfalls of micro-benchmarking)

danielweck commented:
> What we can say for sure is that the optimized version is faster than the current goober one.

Yes, this is consistently observed, in all test runners (Benchmark.js in Node, ESBench and Perf.link in all web browsers).

cristianbote commented:
Totally with you both on this one, the optimised version is overall better. Thank you so much for exploring alternatives @danielweck and @sastan. ✌️

danielweck commented:
ok, i’ll remove the twind benchmark before merge.

cristianbote commented:
@danielweck I would love to have it there. Is there a reason you would want to remove it?

danielweck commented:
i am talking specifically about the Twind entry, not the Goober old/new hashing function benchmark.

cristianbote commented:
Fine by me @danielweck 😃  the more the merrier! Especially when the perf of the twind it's better. That means we have a guideline. 💪 

Thank you so much for all of this! 💯

danielweck commented:
Ha ha :) You beat me to it! In my fork/branch I had just removed the Twind cyrb32 hash and added some comments in the benchmark:

https://github.com/cristianbote/goober/compare/master...danielweck:master

```
// Note: performance comparisons between hash implementations
// produce susprinsingly different results depending on web browsers.
// (see https://github.com/cristianbote/goober/pull/271 where tests are made
// with https://perf.link and https://esbench.com )
// For example, it turns out that Math.imul() -based hashing is
// more performant in Firefox and Safari, than in Chrome.
// That being said, Goober's "improved" hash implementation (see below)
// is consistently faster in all JS runtimes (including Node).
```

Never mind, this is a minor detail :)

danielweck commented:
@sastan I noticed that Beamwind includes a `Math.imul()` polyfill, so I ran some tests to compare Twind's `cyrb32` with native vs. polyfilled `Math.imul()`:

[PERF.LINK](https://perf.link/#eyJpZCI6InQxb3BjbGxqbXNwIiwidGl0bGUiOiJIQVNIISIsImJlZm9yZSI6ImNvbnN0IFNUUiA9IGBhYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3l4eiBBQkNERUZHSElKS0xNTk9QUVJTVFVWV1lYWiAwMTIzNDU2Nzg5YDtcblxuY29uc3QgdHdpbmROYXRpdmUgPSAoc3RyKSA9PiB7XG4gICAgbGV0IGkgPSBzdHIubGVuZ3RoIC0gMSwgb3V0ID0gOTtcbiAgICB3aGlsZSAoaSA%2BPSAwKSBvdXQgPSBNYXRoLmltdWwob3V0IF4gc3RyLmNoYXJDb2RlQXQoaS0tKSwgMHg1ZjM1NjQ5NSk7XG4gICAgY29uc3QgYyA9ICdnbycgKyAoKG91dCBeIChvdXQgPj4%2BIDkpKSA%2BPj4gMCkudG9TdHJpbmcoMzYpO1xufTtcblxuY29uc3QgaW11bCA9ICgob3BBLCBvcEIpID0%2BIHtcbiAgICBvcEIgfD0gMDtcbiAgICBsZXQgcmVzdWx0ID0gKG9wQSAmIDB4MDAzZmZmZmYpICogb3BCO1xuICAgIGlmIChvcEEgJiAweGZmYzAwMDAwIC8qICE9PSAwICovKSByZXN1bHQgKz0gKChvcEEgJiAweGZmYzAwMDAwKSAqIG9wQikgfCAwO1xuICAgIHJldHVybiByZXN1bHQgfCAwO1xufSk7XG5cbmNvbnN0IHR3aW5kUG9seSA9IChzdHIpID0%2BIHtcbiAgICBsZXQgaSA9IHN0ci5sZW5ndGggLSAxLCBvdXQgPSA5O1xuICAgIHdoaWxlIChpID49IDApIG91dCA9IGltdWwob3V0IF4gc3RyLmNoYXJDb2RlQXQoaS0tKSwgMHg1ZjM1NjQ5NSk7XG4gICAgY29uc3QgYyA9ICdnbycgKyAoKG91dCBeIChvdXQgPj4%2BIDkpKSA%2BPj4gMCkudG9TdHJpbmcoMzYpO1xufTsiLCJ0ZXN0cyI6W3sibmFtZSI6Ik1hdGguaW11bCgpIHBvbHlmaWxsIiwiY29kZSI6InR3aW5kUG9seShTVFIpOyIsInJ1bnMiOlsxMDAwLDEwMDAsMTAwMCwyMDAwLDkwMDAsNzAwMDAsODAwMCwyNzAwMCwyNDAwMCwxMDAwLDM5MDAwLDIwMDAsMzcwMDAsNzkwMDAsMTAwMCwxMzAwMCw0NzAwMCw0NDAwMCwxMDAwLDkwMDAsMTAwMCwxMDAwLDE4MDAwLDI4NTAwMCwyMDAwLDkwMDAsNjkwMDAsNDEwMDAsMzQwMDAsNTMwMDAsMzEyMDAwLDEwMDAsODEwMDAsMjAwMCwyMDAwLDEwMDAsMjAwMCwyMjAwMCwyODQwMDAsMTY1MDAwLDEwMDAsMjQwMDAsMTAwMCwyMjYwMDAsMTA0MDAwLDIwMDAsODYwMDAsMjAwMCwxMzAwMCwyMDAwLDEwMDAsMjAwMCwxMDAwLDI0MDAwLDEwMDAsMTAwMCwxMDAwLDE2MDAwLDE5MDAwLDEwMDAsMTAwMCwxOTUwMDAsMjAwMCwzODAwMCwxMDAwLDEwMDAsNDEwMDAsMjAwMCw1MzAwMCwzMTAwMCw0ODAwMCw3OTAwMCwxMDAwLDc2MDAwLDEwMDAsMjM0MDAwLDEwMDAsNjkwMDAsMTAwMCwxMDAwLDMyMDAwLDIwMDAsMTAwMCwxMDAwLDIwMDAsNDkwMDAsODUwMDAsNDEwMDAsMjQyMDAwLDE3MTAwMCwyNDAwMCwxMDAwLDMwMDAwLDcxMDAwLDI1OTAwMCw2MDAwLDM4MDAwLDMwMDAwLDcyMzAwMCwyMDAwXSwib3BzIjo1MDE1MH0seyJuYW1lIjoiTWF0aC5pbXVsKCkgbmF0aXZlIiwiY29kZSI6InR3aW5kTmF0aXZlKFNUUik7IiwicnVucyI6WzAsMTAwMCwyMDAwLDEwMDAwLDI1MDAwLDMzMDAwMCwyNDAwMCw5NTAwMCw4MzAwMCwxMDAwLDEwNTAwMCwxMDAwMCwxMTMwMDAsNDE2MDAwLDEwMDAsNjIwMDAsMTMzMDAwLDEwODAwMCwxMDAwLDI0MDAwLDEwMDAsMjAwMCw1NDAwMCwxMDAwLDEwMDAwLDI1MDAwLDM0MjAwMCwxMDQwMDAsMTExMDAwLDIyOTAwMCwxMTgzMDAwLDEwMDAsNDI3MDAwLDkwMDAsMTAwMDAsMTAwMCwxMDAwMCw3MTAwMCwxMDk3MDAwLDczOTAwMCwxMDAwLDkzMDAwLDIwMDAsOTExMDAwLDc2NDAwMCwyMDAwLDQyMzAwMCw5MDAwLDE5MDAwLDI0MDAwLDEwMDAsMTAwMDAsMTAwMCwzMTUwMDAsMTAwMCwxMDAwLDU1NzAwMCwxMDAwMCwyNDAwMCwyMDAwLDEwMDAsOTg3MDAwLDE2MDAwLDUzMDAwLDkwMDAsMjAwMCwxMTgwMDAsODAwMCwxMzAwMDAsOTkwMDAsMTI5MDAwLDM4NzAwMCwxMDAwLDQ1NzAwMCwxMDAwLDkxNDAwMCwxMDAwLDE2MzAwMCwxMDAwLDIwMDAsMTAxMDAwLDEwMDAwLDEwMDAsMTAwMCwxMDAwMCwxMzYwMDAsNDkxMDAwLDExMzAwMCw5NjEwMDAsODIzMDAwLDg3MDAwLDEwMDAsNjMwMDAsMzI1MDAwLDEwODYwMDAsMTQwMDAsMTEzMDAwLDc5MDAwLDI1NDQwMDAsMTgwMDBdLCJvcHMiOjE5NTk3MH1dLCJ1cGRhdGVkIjoiMjAyMS0wMy0yNlQwODozODowNy4yMzlaIn0%3D)


Unsurprisingly, native `Math.imul()` is 3-4x faster in Firefox and Safari, however it is only 2x faster in Chrome. Interestingly, the `Math.imul()` polyfill consistently runs super fast in Firefox, much less so in Safari, and definitely slower in Chrome. 

Overall, this explains why Twind's `cyrb32` hash routine is slower than Goober's in Chrome (probably explains the poor performance in Node V8 too).

PS: I am getting similar performance ratios with https://esbench.com
